pkg/trie/triedb/codec/node.go

```
@@ -75,19 +104,91 @@ type (
 	// Leaf always contains values
 	Leaf struct {
 		PartialKey []byte
-		Value      NodeValue
+		Value      EncodedValue
 	}
 	// Branch could has or not has values
 	Branch struct {
 		PartialKey []byte
 		Children   [16]MerkleValue
-		Value      NodeValue
+		Value      EncodedValue
 	}
 )
 
-func (Empty) GetPartialKey() []byte    { return nil }
-func (Empty) GetValue() NodeValue      { return nil }
-func (l Leaf) GetPartialKey() []byte   { return l.PartialKey }
-func (l Leaf) GetValue() NodeValue     { return l.Value }
-func (b Branch) GetPartialKey() []byte { return b.PartialKey }
-func (b Branch) GetValue() NodeValue   { return b.Value }
+func (Empty) GetPartialKey() []byte     { return nil }
+func (Empty) GetValue() EncodedValue    { return nil }
+func (l Leaf) GetPartialKey() []byte    { return l.PartialKey }
+func (l Leaf) GetValue() EncodedValue   { return l.Value }
+func (b Branch) GetPartialKey() []byte  { return b.PartialKey }
+func (b Branch) GetValue() EncodedValue { return b.Value }
+
+type NodeKind int
+
+const (
+	LeafNode NodeKind = iota
+	BranchWithoutValue
+	BranchWithValue
+	LeafWithHashedValue
+	BranchWithHashedValue
+)
+
+func EncodeHeader(partialKey []byte, kind NodeKind, writer io.Writer) (err error) {
+	partialKeyLength := len(partialKey)
+	if partialKeyLength > int(maxPartialKeyLength) {
+		panic(fmt.Sprintf("partial key length is too big: %d", partialKeyLength))
```

panic is an attack vector in blockchain systems. Do we absolutely need it here or we can return an error?